gcp_variant_transforms/testing/integration/run_tests_common.py

```diff
@@ -0,0 +1,191 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Common functions and classes that are used by integration tests.
+
+It provides common functions and classes for both run_vcf_to_bq_tests
+(integration test script for vcf_to_bq pipeline) and run_preprocessor_tests
+(integration test script for vcf_to_bq_preprocess pipeline).
+"""
+
+import argparse  # pylint: disable=unused-import
+import json
+import os
+import time
+from typing import Dict, List  # pylint: disable=unused-import
+
+from googleapiclient import discovery
+from oauth2client.client import GoogleCredentials
+
+_DEFAULT_IMAGE_NAME = 'gcr.io/gcp-variant-transforms/gcp-variant-transforms'
+_DEFAULT_ZONES = ['us-west1-b']
+
+
+class TestCaseFailure(Exception):
+  """Exception for failed test cases."""
+  pass
+
+
+class TestRunner(object):
+  """Runs the tests using pipelines API."""
+
+  def __init__(self, tests):
+    # type: (List[Union[PreprocessorTestCase, VcfToBQTestCase]]) -> None
+    self._tests = tests
+    self._service = discovery.build(
+        'genomics',
+        'v1alpha2',
+        credentials=GoogleCredentials.get_application_default())
+    self._operation_names = []  # type: List[str]
+    self._responses = []  # List[Dict]
+
+  def run(self):
+    """Runs all tests."""
+    for test in self._tests:
+      # The following pylint hint is needed because `pipelines` is a method that
+      # is dynamically added to the returned `service` object above. See
+      # `googleapiclient.discovery.Resource._set_service_methods`.
+      # pylint: disable=no-member
+      request = self._service.pipelines().run(body=test.pipeline_api_request)
+      operation_name = request.execute()['name']
+      self._operation_names.append(operation_name)
+    self._wait_for_all_operations_done()
+
+  def _wait_for_all_operations_done(self):
+    """Waits until all operations of `service` are done."""
+    # pylint: disable=no-member
+    operations = self._service.operations()
+    running_operation_names = set()
+    for operation_name in self._operation_names:
+      running_operation_names.add(operation_name)
+    while running_operation_names:
+      time.sleep(10)
+      for operation_name in self._operation_names:
+        if operation_name in running_operation_names:
+          request = operations.get(name=operation_name)
+          response = request.execute()
+          if response['done']:
+            self._handle_failure(response)
+            self._responses.append(response)
+            running_operation_names.remove(operation_name)
+
+  def _handle_failure(self, response):
+    """Raises errors if test case failed."""
+    if 'error' in response:
+      if 'message' in response['error']:
```

Comment:
This is probably fine but just to make sure, now that `print_errors` is removed, please try a case that fails and see if you are happy with the output that it generates, i.e., it is descriptive enough.

Reply:
Well, before it was getting the output from separate processes and then try to extract the stack trace and things like that (hence, `_get_traceback`, etc., which was ugly). I just wanted to make sure with these simplifications it is easy to debug failures. Thanks for checking.